mantidproject/mantid
# Context menu to toggle error bars on plot

Will be wrapped up after ~~#25945~~ is merged

**Description of work.**
Adds context menu to toggle error bars on plot.

Errors for each line can be toggled individually, or all errors can be toggled/shown/hidden.

**To test:**
#### Setup
Script for **8** script plots and 2 workspaces.
```python
# The following line helps with future compatibility with Python 3
# print must now be used as a function, e.g print('Hello','World')
from __future__ import (absolute_import, division, print_function, unicode_literals)

# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *

import matplotlib.pyplot as plt

import numpy as np
ws2d_histo = CreateWorkspace(DataX=[10, 20, 30, 10, 20, 30, 10, 20, 30],
                             DataY=[2, 3, 4, 5, 3, 5],
                             DataE=[1, 2, 3, 4, 1, 1],
                             NSpec=3,
                             Distribution=True,
                             UnitX='Wavelength',
                             VerticalAxisUnit='DeltaE',
                             VerticalAxisValues=[4, 6, 8])

ws2d_1spec = CreateWorkspace(DataX=[10, 20, 30],
                                    DataY=[2, 3],
                                    DataE=[1, 2],
                                    NSpec=1,
                                    Distribution=True,
                                    UnitX='Wavelength',
                                    VerticalAxisUnit='DeltaE',
                                    VerticalAxisValues=[4])
                                    
def do_plots(bonus_kwargs=None):
    if bonus_kwargs is None:
        bonus_kwargs = {}
    plt.figure()
    plt.title('Script Plot without Errors')
    plt.plot([0,15000], [0,15000], label='MyLabel')
    plt.plot([0,15000], [0,14000], label='MyLabel 2')
    plt.legend()

    x = np.linspace(0, 15000, 150)
    plt.figure()
    plt.title('Script Plot with YErr and XErr.{}'.format('' if len(bonus_kwargs) == 0 else " Bonus ARGS: {}".format(bonus_kwargs)))
    plt.errorbar(x, 2*x, yerr=5*np.sqrt(2*x), xerr=2*np.sqrt(2*x), label='MyLabel', **bonus_kwargs)
    plt.legend()

    x = np.linspace(0, 15000, 150)
    plt.figure()
    plt.title('Script Plot with ONLY YErr.{}'.format('' if len(bonus_kwargs) == 0 else " Bonus ARGS: {}".format(bonus_kwargs)))
    plt.errorbar(x, 2*x, yerr=5*np.sqrt(2*x), label='MyLabel', **bonus_kwargs)
    plt.legend()

    x = np.linspace(0, 15000, 150)
    plt.figure()
    plt.title('Script Plot with ONLY XErr.{}'.format('' if len(bonus_kwargs) == 0 else " Bonus ARGS: {}".format(bonus_kwargs)))
    plt.errorbar(x, 2*x, xerr=2*np.sqrt(2*x), label='MyLabel', **bonus_kwargs)
    plt.legend()

do_plots()
do_plots({'capsize':4})
plt.show()
```

#### To Test
- `ws2d_1spec` - check that the error bar for the 1 spectra can be toggled from:
  - Right Click > Error Bars > Show/Hide all errors
  - Right Click > Error Bars > click on the 1 spectra shown
  - Cogwheel for plot options > Curves > Errorbars > check/uncheck `hide errorbars`
- `ws2d_histo`
  - Same as above, but for more than one curve
- Scripted plots
  - Check that Y and X errors are toggled correctly
    - Toggling the errors should change the state of both Y and X errors at the same time
  - Check that the one without errors doesn't provide the context menu at all

- Check that plots can be saved and loaded correctly, before and after toggling the errorbars.

Fixes #25547

Release notes added in `docs/source/release/v4.1.0/ui.rst`.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
